Source/Parallelization/WarpXComm_K.H

```diff
@@ -641,118 +486,181 @@ void warpx_interp_nd_efield_z (int j, int k, int l,
+#endif
+
+    // Auxiliary variables for stencil coefficients
+#ifdef WARPX_USE_PSATD
+    amrex::Real cj = 1.0_rt;
+    amrex::Real ck = 1.0_rt;
+    amrex::Real cl = 1.0_rt;
+    int dj, dk;
+#if (AMREX_SPACEDIM == 3)
+    int dl;
+#endif
+#endif
+
+    // Min and max for interpolation loop along j
+    const int jmin = (sj == 0) ? j - nj + 1 : j;
+    const int jmax = (sj == 0) ? j + nj - 2 : j;
+
+    // Min and max for interpolation loop along k
+    const int kmin = (sk == 0) ? k - nk + 1 : k;
+    const int kmax = (sk == 0) ? k + nk - 2 : k;
+
+    // Min and max for interpolation loop along l
+#if   (AMREX_SPACEDIM == 2)
+    // l = 0 always
+    const int lmin = l;
+    const int lmax = l;
+#elif (AMREX_SPACEDIM == 3)
+    const int lmin = (sl == 0) ? l - nl + 1 : l;
+    const int lmax = (sl == 0) ? l + nl - 2 : l;
+#endif
+
+    amrex::Real res = 0.0_rt;
+
+#ifndef WARPX_USE_PSATD // FDTD (linear interpolation)
+
+    // Example of 1D interpolation from nodal grid to nodal grid:
+    //
+    //         j
+    // --o-----o-----o--  result(j) = f(j)
+    // --o-----o-----o--
+    //  j-1    j    j+1
+    //
+    // Example of 1D interpolation from staggered grid to nodal grid:
+    //
+    //         j
+    // --o-----o-----o--  result(j) = (f(j-1) + f(j)) / 2
+    // -----x-----x-----
+    //     j-1    j
+
+    for (int ll = lmin; ll <= lmax; ll++)
+    {
+        for (int kk = kmin; kk <= kmax; kk++)
+        {
+            for (int jj = jmin; jj <= jmax; jj++)
+            {
+                res += arr_fine(jj,kk,ll);
+            }
+        }
     }
-    Eza(j,k,l) = res;
-#else
-    amrex::Real res = 0.;
-    for (int nz = 1; nz < noz/2+1; nz++) {
-        res += 0.5_rt * stencil_coef_z[nz-1] * (Ezf(j, k, l + nz - 1) + Ezf(j, k, l - nz));
+
+#else // PSATD (finite-order interpolation)
+
+    // Example of 1D interpolation from nodal grid to nodal grid:
+    //
+    //         j
+    // --o-----o-----o--  result(j) = f(j)
+    // --o-----o-----o--
+    //  j-1    j    j+1
+    //
+    // Example of 1D interpolation from staggered grid to nodal grid:
+    //
+    //                     j
+    // --o-----o-----o-----o-----o-----o-----o--  result(j) = c_0 * (f(j-1) + f(j)  ) / 2
+    // -----x-----x-----x-----x-----x-----x-----            + c_1 * (f(j-2) + f(j+1)) / 2
+    //     j-3   j-2   j-1    j    j+1   j+2                + c_2 * (f(j-3) + f(j+2)) / 2
+    //     c_2   c_1   c_0   c_0   c_1   c_2                + ...
+
+    for (int ll = lmin; ll <= lmax; ll++)
+    {
+#if (AMREX_SPACEDIM == 3)
+        dl = (l - ll > 0) ? l - ll - 1 : ll - l;
+        if (sl == 0) cl = stencil_coeffs_z[dl];
+#endif
+        for (int kk = kmin; kk <= kmax; kk++)
+        {
+            dk = (k - kk > 0) ? k - kk - 1 : kk - k;
+#if   (AMREX_SPACEDIM == 2)
+            if (sk == 0) ck = stencil_coeffs_z[dk];
+#elif (AMREX_SPACEDIM == 3)
+            if (sk == 0) ck = stencil_coeffs_y[dk];
+#endif
```

Comment:
As above, this is now done only at the level of the loop over `kk` instead of the innermost loop over `jj`.